### PR TITLE
[Merged by Bors] - fix: fix Q-smul diamond in Complex

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---
The example that has now been added to `Data.Complex.Module` did not commute before, but now that the definition of `qsmul` in the `Field` instance for complex numbers has been changed, it does commute.

I also fixed the statement of `rat_cast_re`, so it's an equality of real numbers now, without a coercion to complex. I also restored the `norm_cast` attribute on `rat_cast_im` and `rat_cast_re` since there is no error, probably because the coercion to `Complex` has been removed.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Complex/Basic.lean [hunk 1]
```diff
@@ -757,11 +757,57 @@ protected theorem mul_inv_cancel {z : ℂ} (h : z ≠ 0) : z * z⁻¹ = 1 := by
...
+  show (Rat.castRec q : ℂ).re = _
+  cases q
+  simp [Rat.castRec, normSq, Rat.mk_eq_divInt, Rat.mkRat_eq_div, div_eq_mul_inv, *]
+#align complex.rat_cast_re Complex.rat_cast_re
+
+/-! ### Field instance and lemmas -/
 
 noncomputable instance : Field ℂ :=
-{ inv := Inv.inv
+{ qsmul := fun n z => ⟨n • z.re - 0 * z.im, n • z.im + 0 * z.re⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Complex/Basic.lean:808]: Please add a note here why we have the weird definition?

Comment (ChrisHughes24) [reply Mathlib/Data/Complex/Basic.lean]: I added a comment, but I actually don't understand the comment to do with `RestrictScalars` on the instance in `Data.Complex.Module`, so I don't fully understand the reason behind this change.

Comment (ChrisHughes24) [reply Mathlib/Data/Complex/Basic.lean]: I should maybe improve that comment as well.

Comment (ChrisHughes24) [reply Mathlib/Data/Complex/Basic.lean]: I just realised I could also change the definition of `Field` to this, and change the `ratCast` def to go through the reals. This also makes the instance diagram commute, but for a more complex reason.

```lean
noncomputable instance : Field ℂ :=
{ ratCast := fun q => (q : ℝ)
  ratCast_mk := fun a b h1 h2 => by
    show ofReal' _ = _
    simp [Rat.cast_mk' a b h1 h2]; rfl
  inv := Inv.inv
  mul_inv_cancel := @Complex.mul_inv_cancel
  inv_zero := Complex.inv_zero }
```

**Comment by j-loreaux** [comment]
@eric-wieser I know you're not a fan of the weird definition of [docs#Module.complexToReal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Complex/Module.html#Module.complexToReal) for nice defeqs, so I thought you may want to see this.

File: Mathlib/Data/Complex/Basic.lean [hunk 2]
```diff
@@ -757,11 +757,60 @@ protected theorem mul_inv_cancel {z : ℂ} (h : z ≠ 0) : z * z⁻¹ = 1 := by
...
+#align complex.rat_cast_re Complex.rat_cast_re
+
+/-! ### Field instance and lemmas -/
 
 noncomputable instance : Field ℂ :=
-{ inv := Inv.inv
+{
+/- We define `qsmul` in this way to ensure it is equal to the more general instance
+`SMul R ℝ → SMul R ℂ` found in `Data.Complex.Module`. -/
+  qsmul := fun n z => ⟨n • z.re - 0 * z.im, n • z.im + 0 * z.re⟩
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Basic.lean:811]: Why not just use
```suggestion
/- We define `qsmul` in this way to ensure it is equal to the more general instance
`SMul R ℝ → SMul R ℂ` found in `Data.Complex.Module`. -/
  qsmul := fun n z => n • z
```
if that instance already exists?

File: Mathlib/Data/Complex/Basic.lean [hunk 3]
```diff
@@ -827,51 +876,11 @@ theorem normSq_div (z w : ℂ) : normSq (z / w) = normSq z / normSq w :=
...
-@[simp, norm_cast]
-theorem int_cast_im (n : ℤ) : (n : ℂ).im = 0 := by rw [← ofReal_int_cast, ofReal_im]
-#align complex.int_cast_im Complex.int_cast_im
-
 @[simp, norm_cast]
 theorem ofReal_rat_cast (n : ℚ) : ((n : ℝ) : ℂ) = (n : ℂ) :=
   map_ratCast ofReal n
 #align complex.of_real_rat_cast Complex.ofReal_rat_cast
 
--- Porting note: removed `norm_cast` attribute because the RHS can't start with `↑`
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Basic.lean:864]: Where did this porting note go?

Comment (ChrisHughes24) [reply Mathlib/Data/Complex/Basic.lean:864]: I restored the `norm_cast` attribute and deleted the note. I think the error may have been something to do with both sides being coerced to a `Complex`.

Comment (ChrisHughes24) [reply Mathlib/Data/Complex/Basic.lean:864]: I restored the `norm_cast` attribute and deleted the note. I think the error may have been something to do with both sides being coerced to a `Complex`.

File: Mathlib/Data/Complex/Basic.lean (hunk 2 above)
Comment (ChrisHughes24) [reply Mathlib/Data/Complex/Basic.lean:811]: The generic `SMul` instance is defined in `Data.Complex.Module` which is not imported

Comment (eric-wieser) [reply Mathlib/Data/Complex/Basic.lean:811]: Can we move the smul instance (and only that insurance) to this file?

File: Mathlib/Data/Complex/Basic.lean [hunk 4]
```diff
@@ -365,22 +365,46 @@ defined in `Data.Complex.Module`. -/
 instance : Nontrivial ℂ :=
   pullback_nonzero re rfl rfl
 
+section SMul
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Basic.lean:368]:
```suggestion
-- porting note: moved from `Module/Data/Complex/Basic.lean`
section SMul
```

File: Mathlib/Data/Complex/Basic.lean [hunk 5]
```diff
@@ -808,8 +833,8 @@ noncomputable instance instField : Field ℂ :=
 {
 /- We define `qsmul` in this way to ensure it is equal to the more general instance
 `SMul R ℝ → SMul R ℂ` found in `Data.Complex.Module`. -/
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Basic.lean]: This comment is now outdated

File: Mathlib/Data/Complex/Basic.lean [hunk 6]
```diff
@@ -406,6 +430,7 @@ instance Complex.addGroupWithOne : AddGroupWithOne ℂ :=
         rfl
     one := 1 }
 
+
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Basic.lean:433]:
```suggestion
```

File: Mathlib/Data/Complex/Module.lean [hunk 7]
```diff
@@ -57,20 +57,7 @@ section
 
 variable [SMul R ℝ]
```
Comment (eric-wieser) [inline Mathlib/Data/Complex/Module.lean:58]: This line and the section its within is now empty

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**